src/sessions/softphone-session-handler.ts

```diff
@@ -22,22 +96,319 @@ export default class SoftphoneSessionHandler extends BaseSessionHandler {
     }
   }
 
+  handleSoftphoneConversationUpdate (
+    update: ConversationUpdate,
+    participant: IConversationParticipantFromEvent,
+    callState: ICallStateFromParticipant,
+    session?: IExtendedMediaSession
+  ) {
+    const conversationId = update.id;
+    const lastConversationUpdate = this.conversations[conversationId] ? { ...this.conversations[conversationId] } : null;
+
+    if (session) {
+      session.pcParticipant = participant as any;
+    }
+
+    this.log('debug', 'about to process conversation event', { session, update, lastConversationUpdate, callState });
+
+    /* if we didn't have a previous update and this one is NOT in a pending ended state, that means we are not responsible for this conversation (another client handled it or we have already emitted the `sessionEnded` for it) */
+    if (!lastConversationUpdate && !this.isPendingState(callState)) {
+      return this.log('debug', 'received a conversation event for a conversation we are not responsible for. not processing', { update, callState }, true);
+    }
+
+    /* update the conversation state */
+    if (!lastConversationUpdate) {
+      this.conversations[conversationId] = {} as any;
+    }
+
+    /* this is just a precaution – session for the conversation should be added in `this.acceptSession()` */
+    if (!this.conversations[conversationId].session) {
+      this.conversations[conversationId].session = session;
+    }
+
+    this.conversations[conversationId].mostRecentCallState = callState;
```

Comment:
NIT: 
```
this.conversations[conversationId] = {
  mostRecentCallState: callState,
  mostRecentUserParticipant: participant,
  ...
}
```

Reply:
The initial reasoning for doing it this way was to keep the reference pointing at the same object and not setting it to a new one every time. But I can't remember if that actually mattered anyway in this file. I'll look over it again and just do a full reassign if it makes sense.

Reply:
you could also do a `Object.assign(...` if you wanted to keep the same ref and just reduce the verbosity.